presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java

```diff
@@ -195,13 +199,23 @@ public PlanNode visitPlan(PlanNode node, RewriteContext<PlanNode> context)
                         filter);
             }
 
-            if (!graph.getAssignments().isPresent()) {
-                return result;
+            if (graph.getAssignments().isPresent()) {
+                result = new ProjectNode(
+                        idAllocator.getNextId(),
+                        result,
+                        Assignments.copyOf(graph.getAssignments().get()));
             }
-            return new ProjectNode(
-                    idAllocator.getNextId(),
-                    result,
-                    Assignments.copyOf(graph.getAssignments().get()));
+
+            if (!result.getOutputSymbols().equals(node.getOutputSymbols())) {
+                // Introduce a projection to constrain the outputs to what was originally expected
```

Comment:
Right, but you still add `ProjectNode` below. Why such `joinNodeWithEnforcedOutput` could break code if it also added projection? As I see result could be any plan node